### PR TITLE
Support union type

### LINKED ISSUE #6261: Support Union types as expected type for JSON data module
**Description:**

$title

**Describe your task(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes: [#6261](https://github.com/ballerina-platform/ballerina-library/issues/6261)

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

File: native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonParser.java [hunk 1]
```diff
@@ -694,31 +688,7 @@ public State transition(StateMachine sm, char[] buff, int i, int count) throws J
...
-                            sm.expectedTypes.push(fieldType);
-
-                            if (!sm.allowDataProjection && fieldType == null)  {
-                                throw DiagnosticLog.error(DiagnosticErrorCode.UNDEFINED_FIELD, jsonFieldName);
-                            }
-                        } else if (sm.expectedTypes.peek() == null) {
-                            sm.expectedTypes.push(null);
-                        }
-                        sm.fieldNameHierarchy.peek().push(jsonFieldName);
+                        JsonCreator.handleFieldName(jsonFieldName, sm);
```
Comment (prakanth97) [inline native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonParser.java]: `handleFieldName` should be present with in the JsonParser

File: native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonParser.java [hunk 2]
```diff
@@ -897,13 +850,14 @@ public State transition(StateMachine sm, char[] buff, int i, int count) throws J
                         state = FIRST_FIELD_READY_STATE;
                     } else if (ch == '[') {
                         state = FIRST_ARRAY_ELEMENT_READY_STATE;
-                        sm.updateNextArrayValue();
+                        // TODO: Check why we did not update the exp type with member Type.
```
Comment (prakanth97) [inline native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonParser.java]: Remove this

File: native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonParser.java [hunk 3]
```diff
@@ -940,11 +894,12 @@ public State transition(StateMachine sm, char[] buff, int i, int count) throws J
                         JsonCreator.updateNextMapValue(sm);
                         state = FIRST_FIELD_READY_STATE;
                     } else if (ch == '[') {
+                        // TODO: Check why we did not update the exp type with member Type.
```
Comment (prakanth97) [inline native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonParser.java]: Remove this

File: ballerina/tests/from_json_string_test.bal [hunk 4]
```diff
@@ -1552,7 +1552,7 @@ isolated function testParseStringNegative3() returns Error? {
 @test:Config
 isolated function testParseStringNegative4() returns Error? {
     string str = string `{
-        name: "John"
+        "name": "John"
```
Comment (SasinduDilshara) [inline ballerina/tests/from_json_string_test.bal:1555]: why we need this change?

**Comment by SasinduDilshara** [comment]
Shall we add a test with never and error

File: ballerina/tests/from_json_string_with_union_types.bal [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+            "a": "1",
+            "b": 2
+        },
+        "field2": {
+            "a": "3",
+            "b": 4
+        }
+    }`;
+    RecB val = check parseString(jsonStr);
+    test:assertEquals(val.field1, {a: "1", b: 2});
```
Comment (LakshanWeerasinghe) [inline ballerina/tests/from_json_string_with_union_types.bal:65]: Lets test the type of each field as well.

File: ballerina/tests/from_json_string_test.bal (hunk 4 above)
Comment (prakanth97) [reply ballerina/tests/from_json_string_test.bal:1555]: Json Object key should be present within `""`

**Comment by prakanth97** [comment]
> Shall we add a test with never and error

Since both of these are not a subtype of anydata, we will get compile-time error.

**Review by SasinduDilshara — DISMISSED** [review]

**Comment by SasinduDilshara** [comment]
Shall we add tests with a scenario that, union type as a expected type and it match not with source

**Comment by SasinduDilshara** [comment]
Shall we add a test with union of singleton types

**Comment by prakanth97** [comment]
> Shall we add a test with union of singleton types

Already we have test for that https://github.com/ballerina-platform/module-ballerina-data.jsondata/blob/main/ballerina/tests/from_json_string_test.bal#L1051

**Review by SasinduDilshara — APPROVED** [review]

**Pull request merged**